File: cpp/rec_splitter.cpp
// Include the header-only libcluon file.
#include "cluon-complete.hpp"

// Include our self-contained message definitions.
#include "example.hpp"
#include "opendlv-standard-message-set.hpp"

// We will use iostream to print the received message.
#include <iostream>

#include <string>
#include <fstream>
#include <dirent.h>

#include <sys/types.h>
#include <sys/stat.h>

// Function prototypes 
void writeMessageToFile(std::ofstream &fd, cluon::data::Envelope);
int fileType(std::string);
int cutDirectory(std::string, double, double);
int cutFile(std::string, double, double);
bool updateGPXString(std::string&, cluon::data::Envelope, double, bool);

int main(int argc, char **argv) {

  double start_time_relative, time_interval;

  if (argc < 2) {
    std::cerr << "Use: " << argv[0] << " FILENAME [START_TIME] [TIME_INTERVAL]" << std::endl;
    exit(-1);
  }
  else if(argc == 2){
    // Default values if only filename is given
    time_interval = 5.0;
    start_time_relative = 100.0;
  }

  // Non-error tolerant argument parsing
  else if(argc > 2 && argc < 5){
    start_time_relative = std::stod(argv[2]);
    time_interval = std::stod(argv[3]);
  }
  else {
    std::cerr << "Use: " << argv[0] << " FILENAME [START_TIME] [TIME_INTERVAL]" << std::endl;
    exit(-1);
  }

  //time_interval = 5.0;
  //start_time_relative = 0.0;

  std::string filename_in = argv[1];
  //std::string filename_in = "testdir";
  int type = fileType(filename_in);

  // If directory, cut all eligible files into the given interval
  if (type == 0){
    cutDirectory(filename_in, start_time_relative, time_interval);
  }
  // If filename_in is a file then cut it
  else if (type == 1){
    cutFile(filename_in, start_time_relative, time_interval);
  }
  else if (type == 2){
    std::cout << "interpreted filename as invalid rec file" << std::endl;
  }
  return 0;
}


// Write envelope data to file
void writeMessageToFile(std::ofstream &fd, cluon::data::Envelope env){
  std::string data = cluon::serializeEnvelope(std::move(env));
  fd.write(&data[0], data.length());
}

// Determines if given filename is a .rec-file or a directory
// Also checks if it is the output from a previous split
// Returns 0 if filename refers to a directory
// 1 if it is a valid .rec-file
// 2 if it is an invalid .rec-file
// NOTE: This function is allergic to filenames that contains the string
// '_rec'
int fileType(std::string filename){
  char *current_token = strtok(&filename[0], "_.");

  if (current_token != NULL){
    while(current_token != NULL){

      if (strcmp(current_token, "out") == 0){
        return 2;
      }
      else if (strcmp(current_token, "rec") == 0){
        return 1;
      }
      current_token = strtok(NULL, "_.");
    }
  }
  return 0;
}

bool updateGPXString(std::string& gpxString, cluon::data::Envelope env, double timestamp, std::string file, bool firstGPS) {
  
  // Unpack the content...
  auto msg = cluon::extractMessage<opendlv::logic::sensation::Geolocation>(std::move(env));

  // ...and access the data fields.
  // Create GPX file.
  // Change into timestamp into ISO8601
  time_t curr_time = (time_t)timestamp;
  char buf[sizeof "2011-10-08T07:07:09Z"];
  strftime(buf, sizeof buf, "%FT%TZ", gmtime(&curr_time));

  // Fetch geolocations
  std::string lat = std::to_string(msg.latitude());
  std::string lon = std::to_string(msg.longitude());
  std::string ele = std::to_string(msg.altitude());

  // Build first input into GPX file.
  // example from:  
  // https://wiki.openstreetmap.org/wiki/GPX
  if (firstGPS){
    gpxString += "<?xml version=\"1.0\" encoding=\"UTF-8\"?> \n";
    gpxString += "<gpx version=\"1.0\"> \n";
    gpxString += "  <name>" + file + "</name>\n";
    gpxString += "  <trk><name>" + file + "</name><number>1</number><trkseg>\n";
    
    firstGPS = false;
  }
  
  gpxString += "    <trkpt lat=\"" + lat + "\" lon=\"" + lon + "\"><ele>" + ele + "</ele><time>" + buf + "</time></trkpt>\n";

  return firstGPS;
}
// Cut file 'filename_in' and put output in 'filename_in_out'
int cutFile(std::string filename_in, double start_time_relative, double time_interval){
  std::string tmpstr;
  tmpstr += filename_in;

  
  //std::cout << filename_in << std::endl;

  std::size_t found = filename_in.find_last_of("/\\");
  std::string path =  filename_in.substr(0,found);
  std::string file = filename_in.substr(found+1);
  std::string file_out = strcat(strtok(&file[0], "."), "_out.rec");
  std::string filename_out = path + "_out/" + file_out;

  std::ofstream f_out(filename_out, std::ios::out | std::ios::binary);



  if (!f_out){
    std::cout << "Cannot open output file! " << std::endl;
    exit(-1);
  }

  bool firstEnvelope = true;
  bool firstGPS = true;
  double timestamp, start_time, stop_time;

  // We will use cluon::Player to parse and process a .rec file.
  constexpr bool AUTO_REWIND{false};
  constexpr bool THREADING{false};
  //cluon::Player player(std::string(argv[1]), AUTO_REWIND, THREADING);
  cluon::Player player(filename_in, AUTO_REWIND, THREADING);
  int i = 0;
  std::string gpxString = "";
  // Now, we simply loop over the entries.
  while (player.hasMoreData()) {
    auto entry = player.getNextEnvelopeToBeReplayed();
    // The .first field indicates whether we have a valid entry.
    if (entry.first) {
      // Get the Envelope with the meta-information.
      cluon::data::Envelope env = entry.second;

      // Get timestamp
      timestamp = env.sampleTimeStamp().seconds() + 1e-6*env.sampleTimeStamp().microseconds();

     
      // Set timestamp of first envelope to filter start time
      if (firstEnvelope){

        start_time = timestamp + start_time_relative;
        stop_time = start_time + time_interval;

        firstEnvelope = false;
      }

      if (timestamp >= start_time && timestamp <= stop_time){

        // Check whether it is of type GeodeticWgs84Reading.
        if (timestamp >= start_time + i*time_interval/100) {
          if (env.dataType() == opendlv::logic::sensation::Geolocation::ID()) {
            firstGPS = updateGPXString(gpxString, env, timestamp, file, firstGPS);
            i++;
          }
        }

        writeMessageToFile(f_out, env);
      }
      else if (timestamp > stop_time){
        break;
      }

    }
  }
  if (firstGPS == false){
    gpxString += "  </trkseg></trk>\n";
    gpxString += "</gpx>";
    std::string gps_out = path + "_out/gps.gpx";
    std::ofstream f_out2(gps_out, std::ios::out | std::ios::binary);
    f_out2.write(&gpxString[0], gpxString.length());
    f_out2.close();
  }
  f_out.close();
  return 0;
}


int cutDirectory(std::string dir_name, double start_time_relative, double time_interval){
  std::string tmpstr = "";
  tmpstr += dir_name;
  tmpstr += '/';
  std::string newFolder = "";
  newFolder += dir_name;
  newFolder += "_out/";
  if (mkdir(newFolder.c_str(), S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH) == -1)
  {
      if( errno == EEXIST ) {
        // alredy exists
      } else {
        // something else
          std::cout << "cannot create sessionnamefolder error:" << strerror(errno) << std::endl;
      }
  }
  
  if (auto dir = opendir(&dir_name[0])){
    while (auto f = readdir(dir)){

      if (fileType(f->d_name) == 1){
        //std::cout << strcat(&tmpstr[0],f->d_name) << std::endl;
        cutFile(tmpstr + f->d_name, start_time_relative, time_interval);
      }
    }
  }
}

File: python3/rec_splitter.py
#!/usr/bin/python3

import sys
from os import listdir
import struct

# Default messages from libcluon providing Envelope and TimeStamp
import cluonDataStructures_pb2
# OpenDLV Standard Message Set
import opendlv_standard_message_set_v0_9_9_pb2
# Messages from individual message set
import example_pb2
# Parse arguments from command line
import argparse

parser = argparse.ArgumentParser(description='Filter a rec file based on time')
parser.add_argument('-r', metavar='recording', default='test.rec')
parser.add_argument('-s', metavar='start_time', default=0)
parser.add_argument('-i', metavar='time_interval', default=1)
args = parser.parse_args()

################################################################################
# Write message to file 
def writeMessageToFile(fd, header, payload):
    fd.write(header)
    fd.write(payload)


################################################################################
# Filter messages in .rec file based on sample timestamp
def filterFile(filename_in, filename_out, start_time_relative, time_window):
    assert filename_in != filename_out, 'Input and output rec files are the same!'
    first_envelope = True

    with open(filename_in, "rb") as f_in:
        with open(filename_out, 'wb') as f_out:
            buf = b""
            bytesRead = 0
            expectedBytes = 0
            LENGTH_ENVELOPE_HEADER = 5
            consumedEnvelopeHeader = False


            byte = f_in.read(1)
            while byte != "":
                buf =  b"".join([buf, byte])
                bytesRead = bytesRead + 1

                if consumedEnvelopeHeader:
                    if len(buf) >= expectedBytes:
                        envelope = cluonDataStructures_pb2.cluon_data_Envelope()
                        envelope.ParseFromString(buf)

                        # Convert timestamp from 2 ints to a float
                        timestamp = envelope.sampleTimeStamp.seconds + 1e-6*envelope.sampleTimeStamp.microseconds

                        # Extract timestamp from first envelope
                        if first_envelope:
                            first_envelope = False
                            start_time = timestamp
                            stop_time = start_time + float(time_window)
                        
                        # Check whether current timestamp is within limits 
                        if timestamp >= start_time+start_time_relative and timestamp <= stop_time:
                            writeMessageToFile(f_out, header, buf)
                        elif timestamp > stop_time:
                            break

                        # Start over and read next container.
                        consumedEnvelopeHeader = False
                        buf = buf[expectedBytes:]
                        expectedBytes = 0

                if not consumedEnvelopeHeader:
                    if len(buf) >= LENGTH_ENVELOPE_HEADER:
                        consumedEnvelopeHeader = True
                        byte0 = buf[0]
                        byte1 = buf[1]

                        # Check for Envelope header.
                        if byte0 == 0x0D and byte1 == 0xA4:
                            v = struct.unpack('<L', buf[1:5]) # Read uint32_t and convert to little endian.
                            expectedBytes = v[0] >> 8 # The second byte belongs to the header of an Envelope.
                            header = buf
                            buf = buf[expectedBytes:] # Remove header.
                        else:
                            print("Failed to consume header from Envelope.")

                # Read next byte.
                byte = f_in.read(1)

                # End processing at file's end.
                if not byte:
                    break


################################################################################
# Does time filtering on all files in a directory
def filterAllFilesInDir(dir_path, start_time_relative, time_window):
    files = listdir(dir_path)
    for e in files:
        name_type = e.split('.')
        
        # Only filter .rec-files
        if len(name_type) == 2 :

            # Check if file is already a splitted version of another
            # then don't split
            if name_type[0].find('_out') < 0:

                # Only split files with .rec ending
                if name_type[1] == 'rec':
                    filterFile(e, name_type[0] + '_out.' + name_type[1], start_time_relative,\
                        time_window)

################################################################################
# Main entry point.
# if len(sys.argv) != 2:
    # print("Display Envelopes captured from OpenDLV.")
    # print("  Usage: %s example.rec" % (str(sys.argv[0])))
    # sys.exit()
start_time_relative = float(args.s)
time_window = float(args.i)
filename_in = args.r

splitted_filename = filename_in.split('.')

# Only process if file is on format x.rec
if len(splitted_filename) == 2 and splitted_filename[-1] == 'rec': 

    filename_out = splitted_filename[0] + '_out.rec'
    filterFile(filename_in, filename_out, start_time_relative, time_window)

# Else interpret recording name as a directory
else:

    filterAllFilesInDir(filename_in, start_time_relative, time_window)